### ISSUE TITLE
All mined blocks are stale

### ISSUE THREAD

**Issue by solardiz** [zcash/zcash#1055]
All of the blocks that `zcashd -gen` manages to mine for me, which happens every few minutes, turn out to be stale:

```
2016-06-25 12:31:14 CreateNewBlock(): total size 1000
2016-06-25 12:31:14 Running ZcashMiner with 1 transactions in block (274 bytes)
2016-06-25 12:31:34 UpdateTip: new best=57fb9e387420d8e936a19237b866dc24412f4c5d73c00cde1ac0d73d99de205e  height=7674  log2_work=17.093479  tx=7690  date=2016-06-25 13:25:42 progress=1.000000  cache=0.0MiB(3tx)
2016-06-25 12:32:45 UpdateTip: new best=24c9f7c5d58babc029b4d6824fb9f8694a12561938eb0c7386920e5973d6b9e7  height=7675  log2_work=17.0935  tx=7691  date=2016-06-25 12:31:46 progress=1.000000  cache=0.0MiB(4tx)
2016-06-25 12:33:54 UpdateTip: new best=42498a8926f057a6e53ea3868b80643e102aded4e003e99e670b2cf09438d15b  height=7676  log2_work=17.093521  tx=7692  date=2016-06-25 12:32:51 progress=1.000000  cache=0.0MiB(5tx)
2016-06-25 12:34:03 ZcashMiner:
2016-06-25 12:34:03 proof-of-work found  
  hash: 50cefaa3ab8a2d0e3758870d6265ef3e3fbcca9f207ad7b1c47fcbbbf700e9ff  
target: 75c28e0000000000000000000000000000000000000000000000000000000000
2016-06-25 12:34:03 CBlock(hash=50cefaa3ab8a2d0e3758870d6265ef3e3fbcca9f207ad7b1c47fcbbbf700e9ff, ver=4, hashPrevBlock=425655fe7074053901bb0e4d14873e4649a87cafb9bc7513ceab3fd94595e65e, hashMerkleRoot=d0e789d8896321951f94bb78ad912620607a848e6be5245a5f78feab28986cf1, hashReserved=0000000000000000000000000000000000000000000000000000000000000000, nTime=1466857874, nBits=2075c28e, nNonce=0000000000000000000000000000000000000000000000000000000000000001, vtx=1)
  CTransaction(hash=d0e789d889, ver=1, vin.size=1, vout.size=2, nLockTime=0)
    CTxIn(COutPoint(0000000000, 4294967295), coinbase 02fa1d0101)
    CTxOut(nValue=3.83700000, scriptPubKey=037c5fa12ad4b93feacc3f1ace9bc3)
    CTxOut(nValue=0.95925000, scriptPubKey=OP_HASH160 6708e6670db0b950dac)

  vMerkleTree:  d0e789d8896321951f94bb78ad912620607a848e6be5245a5f78feab28986cf1

2016-06-25 12:34:03 generated 3.837
2016-06-25 12:34:03 ERROR: ZcashMiner: generated block is stale
```

This might very well be user error or such, but while we're at this early stage of the project I hope it's OK to use GitHub issues for tracking of things like this. Please feel free to close this issue if it's in fact not with the code.

**Comment by str4d** [comment]
The cause of this is that the Equihash solver is currently not interruptible, and takes a minimum of ~50s to solve [on our benchmarks](https://speed.z.cash/timeline/?exe=1&base=1%2B43&ben=time+solveequihash&env=1&revs=10&equid=off&quarts=on&extr=on). Thus if a new block arrives during that interval, it is possible for the solver to finish and find a solution that creates a valid block, only for it to be already stale.

The fix for this requires reworking the solver to be interruptible, which will be done while making it multithreaded (as part of #857).

**Comment by solardiz** [comment]
> Thus if a new block arrives during that interval, it is possible for the solver to finish and find a solution that creates a valid block, only for it to be already stale.

Yes, but is it expected that this would be happening literally all the time for me so far? For 101 out of 101 blocks mined here so far. Maybe someone has a CPU with a higher turbo clock rate and that person only runs 1 thread, but isn't different miners' nonce selection randomized separately, meaning that even a moderately slower miner should have a non-negligible chance?

**Comment by str4d** [comment]
> is it expected that this would be happening literally all the time for me so far? For 101 out of 101 blocks mined here so far.

We've had at least one user say that they have more success mining non-stale blocks with a single thread than several. That would tend to suggest that the solver time per-thread is increasing as the number of threads increases (implying memory contention), which I wouldn't be surprised about currently.

> isn't different miners' nonce selection randomized separately, meaning that even a moderately slower miner should have a non-negligible chance?

Currently they aren't (see #1033), but the `extraNonce` is set per-thread IIRC so the threads should still be doing different work. The main issue here IMHO is that the time per solver thread is around 40% of the target block interval, so there are a) very few solver runs in each interval, and b) only two that complete inside the interval.

**Comment by jl777** [comment]
if you have a relatively slow system, then your odds of finding a block goes down nonlinearly as compared to benchmark speed differences

there are some miners running with 40 seconds benchmark times, mine runs at 80 seconds, with a block being ~150 seconds, then I usually get 1 shot at winning a block. the 40 second machine gets 3 or even 4. so instead of 2x, it can get 3x to 4x the blocks on a 2x faster system.

this effect is independent of interrupting mining threads, since the mining has to restart at the beginning of each new block. As an example if your system takes on average longer than the blocktime you simply never get a chance to submit a block and so never have a chance to mine the block, ie 100% stale.

but it gets worse...

with the diff adjustments every block, the diff levels are increased so that it takes the 2.5 minutes for each block, so that means at some multiple of the fastest system, things will converge to a state where such systems have zero chance to submit a non-stale block. 

Alternatively, if current diff hasnt adjusted yet, then the faster machines are simply finding blocks faster than slow systems can generate even a single set of solutions.

**Comment by solardiz** [comment]
I patched extraNonce initial values and updates to stagger them a bit (not necessarily optimally, and I don't know if this worked at all), and I left this running with only 2 threads (hopefully, reaching max turbo on this dual-socket machine) for many hours, and now I have:

```
$ fgrep -c 'proof-of-work found' ~/.zcash/testnet3/debug.log ; fgrep -c 'generated block is stale' ~/.zcash/testnet3/debug.log
637
626
```

(200+ of these happened before my change/restart above, though), and:

```
    "balance" : 35.91400000,
```

So it looks like 11 blocks were accepted.

The machine is 2xE5-2670, 8xDDR3-1600. It's good for high concurrency, but is in fact slower than modern desktop machines in terms of per-thread performance (max turbo clock rate is 3.3 GHz, whereas some modern desktop CPUs exceed 4 GHz). Also, I think the build does not use maximum compiler optimization, but that's probably the same for others?

Checking debug.log for last 3 hours (exactly), I see 176 instances of "Running ZcashMiner" and 104 instances of "proof-of-work found". That's 61 and 104 seconds, respectively, between these events on average.

**Comment by solardiz** [comment]
> If current diff hasnt adjusted yet, then the faster machines are simply finding blocks faster than slow systems can generate even a single set of solutions.

This appeared to be the case: during a 3-hour interval, blocks were found on average every 72 seconds, instead of the target 150 seconds. Moreover, surprisingly few blocks took significantly longer than that, thereby not giving my system much chance (which probably needed 61*2=122 seconds? sounds rather high indeed - need to optimize that build or something). I guess the reason for this unusual distribution of block times (all being too close to average) is that many other miners are experiencing the same granularity problem. So if one fails to find a solution from its first try, this is immediately followed by many competitors completing their first tries, and one of those very likely succeeds.

**Comment by jl777** [comment]
I am seeing only about a 20% variance in times to find a single solution, so unlike normal hashing where you have a really, really small chance a large number of times, here we have very few chances and the granularity is a big problem for systems that are at the borderline of 1/2 blocktime. a bit longer and 50% of the hashrate is wasted. for systems on the border of a full blocktime, most all the hashrate is wasted. only when a faster than average run happens when no other miner has already found a block will a non-stale block be found.

now it is a bit misleading with all the stales as I think virtually all blocks each miner will find a potential solution. So it is partly psychological. ie, instead of silence, we get "you found a block!, nope, sorry it was stale".

I think as long as more hashrate is being added, the average blocktime will be below the target time and making the majority of nodes 100% stale.

**Comment by daira** [comment]
If the formulae in the Equihash paper are correct, then the time for k=8, n=180 is only a sixth of that for k=3, n=96; the memory usage is similar (~0.83 times as much), and the solution size is still practical (672 bytes for the indices). Another example is k=10, n=198 which is 19.2 times faster and has a solution size of 2432 bytes for the indices. Both of these require a generalization to allow n/(k+1) not to be a multiple of 8, but I see no obstacle to doing that.

**Comment by daira** [comment]
It seems as thought the Equihash authors might have been aassuming a much tighter constraint on solution size than we actually have. (Although this is confusing given their Argon2 PoW proposal which has much larger solutions.)

**Comment by daira** [comment]
Also, let's make sure we're not conflating issues; any problem with the difficulty not adjusting quickly or accurately enough needs a separate ticket.

**Comment by daira** [comment]
k=12, n=208 has a memory target 0.8 times that of k=3, n=96, and should be 64 times faster. It has a solution size of 8704 bytes which is a little on the large size, but still very feasible. Also, n/(k+1) = 16 so it should work with the existing code. I'll try it now.

**Comment by daira** [comment]
It looks as though for these parameters the "optimized" algorithm is taking more memory than expected (and than the unoptimized one?) It's late and I'm tired, so I could easily have made a mistake; I'll investigate further tomorrow.

**Comment by daira** [comment]
Hmm, seems like I was running into problems due to the testnet chain (or the regtest genesis block if I use regtest) not having a valid proof of work. I don't know why that caused a memory allocation error. @str4d, how do I test just the PoW, without needing a valid chain?

**Comment by str4d** [comment]
To test the solver on its own, user the performance benchmark tool. To exercise the full PoW on a fake chain, try my [standalone miner](https://github.com/str4d/zcash/tree/standalone-miner).

**Comment by solardiz** [comment]
> To test the solver on its own, user the performance benchmark tool.

Please excuse my ignorance, but where do I find it?

> To exercise the full PoW on a fake chain, try my standalone miner.

Thanks. Built it with `-O2 -momit-leaf-frame-pointer` (should still allow for debugging, including backtrace) instead of `-Og`, getting these speeds on the same machine as I had mentioned above:

```
Running Equihash solver with nNonce = 0000000000000000000000000000000000000000000000000000000000000000
Solutions: 3
ZcashMiner:
proof-of-work found
prevHash: 0000000000000000000000000000000000000000000000000000000000000000
    hash: 52b433e36b254b661aac5b49cd3d7dd610c54bfa73cfc7fc8e1652974fa899d8
  target: 7fffff0000000000000000000000000000000000000000000000000000000000
duration: 68
Running Equihash solver with nNonce = 0000000000000000000000000000000000000000000000000000000000000000
Solutions: 4
ZcashMiner:
proof-of-work found
prevHash: 52b433e36b254b661aac5b49cd3d7dd610c54bfa73cfc7fc8e1652974fa899d8
    hash: 29da430ad14a388fe5a71a1feb783e529b9150db93fdb4f0838a8952d7a6e698
  target: 7fffff0000000000000000000000000000000000000000000000000000000000
duration: 70
Running Equihash solver with nNonce = 0000000000000000000000000000000000000000000000000000000000000000
Solutions: 2
Running Equihash solver with nNonce = 0000000000000000000000000000000000000000000000000000000000000001
Solutions: 4
ZcashMiner:
proof-of-work found
prevHash: 29da430ad14a388fe5a71a1feb783e529b9150db93fdb4f0838a8952d7a6e698
    hash: 2175fd33bc81c7091c8cf6b9897a8ce3442d29ee19ac442941beb385da7855ad
  target: 7fffff0000000000000000000000000000000000000000000000000000000000
duration: 136
```

```
Running Equihash solver with nNonce = 0000000000000000000000000000000000000000000000000000000000000002
Solutions: 1
ZcashMiner:
proof-of-work found
prevHash: 04c9d0b48fdadbc78670384ce6ca962f2961faebbbebf586359712eb9609dbaa
    hash: 454d0f6287e0179f38ed5e581f03feee23ab7b48d8a93666ab6bfd5c9c259dbe
  target: 7fffff0000000000000000000000000000000000000000000000000000000000
duration: 63
```

So one of the reasons why the granularity manifests itself so badly is that usually there are multiple solutions at a time, making it very likely that at least one will be below the current easy target, but only after all are obtained. Does Equihash arrive at them all roughly at the same time, or are some potentially available for checking significantly sooner?

BTW, maybe `standaloneminer` should print partial duration between solutions even when target was not yet reached.

**Comment by str4d** [comment]
> Please excuse my ignorance, but where do I find it?

`./qa/zcash/performance-benchmarks.sh time solveequihash N` where N is the number of times in a row you want to run it. Each run is a plain Equihash solver run with no regard to the exact solutions returned.

> So one of the reasons why the granularity manifests itself so badly is that usually there are multiple solutions at a time, making it very likely that at least one will be below the current easy target, but only after all are obtained. Does Equihash arrive at them all roughly at the same time, or are some potentially available for checking significantly sooner?

If the indices are not being truncated, then all solutions are obtained at essentially the same time (within the time it takes to find collisions within the last list, which is a few seconds as it is single-pass). If the indices are being truncated and invalid partials need to be eliminated, then it is possible that some valid solutions might be found before others, depending on how many partials have been found. In the worst case, there might be several hundred partials and no valid solutions. In the next worst, there might be several hundred partials and only one or two valid solutions that were found very early on.

A refactor of the solver could test valid solutions as soon as they are found, which might shave 5-20s off the time for finding a valid block if the miner is lucky. But it is still in the miner's best interests to wait until they have eliminated all partials before continuing to the next round, although perhaps it might be profitable to drop all remaining partials once e.g. for valid solutions have been found (trading off the remaining elimination time against the probability of finding another solution that happens to be the only usable one from that solver run).

**Comment by str4d** [comment]
> BTW, maybe `standaloneminer` should print partial duration between solutions even when target was not yet reached.

I will add a timer between finding Equihash solutions tonight (separate to the existing duration timer which is the cumulative time).

**Comment by solardiz** [comment]
> To exercise the full PoW on a fake chain, try my standalone miner.

Related: https://github.com/ebfull/zcash/tree/tromp-standalone-equihash which has this commit:

```
commit de619884c9921abbee4be45facffbd270d8db8ef
Author: Sean Bowe <ewillbefull@gmail.com>
Date:   Wed Apr 20 20:27:47 2016 -0600

    Standalone equihash miner binary, just creates a solution.
```

adding equihash_solver_standalone.cpp, but @str4d's is more elaborate.

**Comment by str4d** [comment]
That's the commit I used to figure out how to add a new executable. The contents of that branch is long out of date, replaced by the code in `src/zcbenchmark.cpp` (which is run via `qa/zcash/performance-benchmarks.sh`).

**Comment by zapv0lt** [comment]
I think that the stale blocks are caused by the current miner not being interrupted/restarted when a new block is found by someone else.  At low difficulty levels, you are likely to mine many blocks but there also is a high chance someone else will beat you to it and submit the proof-of-work before you at some point during your ~50s solution time.  Your miner will not be interrupted at this point, and it will carry on running, find a valid proof-of-work (quite likely due to low difficulty) and then discover that it is "stale".

**Comment by str4d** [comment]
What I'd like to do is make the interrupts asynchronous. Instead of having to check all conditions every round, I'd like some other thread or code to update the miner state when a rebuild is necessary, and then the miner only needs to check a single global flag. This will also probably be necessary to handle the potentially nested sets of threads, which boost::thread_interrupt isn't really suitable for.

**Comment by jl777** [comment]
I hate to sound like a broken record, but C is better suited when you need to deal with things like this, but maybe a workaround would be to spawn a new thread for each mining task and just kill it if it has become stale. Maybe C++ can handle that?

**Comment by daira** [comment]
I don't see how this would differ significantly between C and C++. We may well do a plain-C miner, but none of the immediate issues should or need to block on that.

**Comment by jl777** [comment]
I am no expert on C++ error handling, but posts like http://250bpm.com/blog:4  "The decoupling between raising of the exception and handling it, that makes avoiding failures so easy in C++, makes it virtually impossible to guarantee that the program never runs info undefined behaviour"

indicates there are some potential issues, especially with error handling. Probably this does not apply here, but I am not sure

**Comment by zookozcash** [comment]
This was originally in this milestone but, @str4d is prioritizing (a) thesis defense, and (b) finishing up some work for the Briar Project, so we're kicking this out of this Milestone.
